File: tau_core/web/js/flutter_sound/flutter_sound.js
/*
 * Copyright 2018, 2019, 2020 Dooboolab.
 *
 * This file is part of Flutter-Sound.
 *
 * Flutter-Sound is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License version 3 (LGPL-V3), as published by
 * the Free Software Foundation.
 *
 * Flutter-Sound is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with Flutter-Sound.  If not, see <https://www.gnu.org/licenses/>.
 */




const codec =
                {
                        defaultCodec:   'defaultCodec',
                        aacADTS:        'aacADTS',
                        opusOGG:        'opusOGG',
                        opusCAF:        'opusCAF',
                        mp3:            'mp3',
                        vorbisOGG:      'vorbisOGG',
                        pcm16:          'pcm16',
                        pcm16WAV:       'pcm16WAV',
                        pcm16AIFF:      'pcm16AIFF',
                        pcm16CAF:       'pcm16CAF',
                        flac:           'flac',
                        aacMP4:         'aacMP4',
                        amrNB:          'amrNB',
                        amrWB:          'amrWB',
                        pcm8:           'pcm8',
                        pcmFloat32:     'pcmFloat32',
                        pcmWebM:        'pcmWebM',
                        opusWebM:       'opusWebM',
                        vorbisWebM:     'vorbisWebM',
                };


const tabCodec =
                [
                        codec.opusWebM, // codec.defaultCodec,
                        codec.aacADTS,
                        codec.opusOGG,
                        codec.opusCAF,
                        codec.mp3,
                        codec.vorbisOGG,
                        codec.pcm16,
                        codec.pcm16WAV,
                        codec.pcm16AIFF,
                        codec.pcm16CAF,
                        codec.flac,
                        codec.aacMP4,
                        codec.amrNB,
                        codec.amrWB,
                        codec.pcm8,
                        codec.pcmFloat32,
                        codec.pcmWebM,
                        codec.opusWebM,
                        codec.vorbisWebM,
                ];

const mime_types =
                [
                        'audio/webm\;codecs=opus', // defaultCodec,
                        'audio/aac', // aacADTS,
                        'audio/opus\;codecs=opus', // opusOGG,
                        'audio/x-caf', // opusCAF,
                        'audio/mp3', // mp3,
                        'audio/ogg\;codecs=vorbis', // vorbisOGG,
                        'audio/pcm', // pcm16,
                        'audio/wav\;codecs=1', // pcm16WAV,
                        'audio/aiff', // pcm16AIFF,
                        'audio/x-caf', // pcm16CAF,
                        'audio/x-flac', // flac,
                        'audio/mp4', // aacMP4,
                        'audio/AMR', // amrNB,
                        'audio/AMR-WB', // amrWB,
                        'audio/pcm', // pcm8,
                        'audio/pcm', // pcmFloat32,
                        'audio/webm\;codecs=pcm', // pcmWebM,
                        'audio/webm\;codecs=opus', // opusWebM,
                        'audio/webm\;codecs=vorbis', // vorbisWebM
                ];

const tabFormat =
                [
                        'opus', // defaultCodec,
                        'aac', // aacADTS,
                        'opus', // opusOGG,
                        'caf', // opusCAF,
                        'mp3', // mp3,
                        'vorbis', // vorbisOGG,
                        '', // pcm16,
                        'wav', // pcm16WAV,
                        'aiff', // pcm16AIFF,
                        'caf', // pcm16CAF,
                        'flac', // flac,
                        'mp4', // aacMP4,
                        'AMR', // amrNB,
                        'AMR-WB', // amrWB,
                        '', // pcm8,
                        '', // pcmFloat32,
                        'pcm', // pcmWebM,
                        'opus', // opusWebM,
                        'webm', // vorbisWebM

                ];


var instanceNumber = 0;
var lastUrl = '';



function getRecordURL( aPath,)
{
        console.log('getRecordUrl: ' + aPath);
        var path ;
        var myStorage;
        if ((aPath == null) || (aPath == ''))
        {
                console.log('lastUrl = ' + lastUrl);
                path = lastUrl;
        } else
        {
                path =  aPath;

        }
        if (path.substring(0,1) == '/')
        {
                myStorage = window.localStorage;
                console.log('localStorage');
        } else
        {
                myStorage = window.sessionStorage;
                console.log('sessionStorage');
        }

        var url = myStorage.getItem(path);
        console.log('<--- getRecordUrl ( ' + path  + ' ) : ' + url);
        return url;
}

File: lib/public/flutter_sound.dart
/*
 * Copyright 2018, 2019, 2020 Dooboolab.
 * Copyright 2021, 2022, 2023, 2024 Canardoux.
 *
 * This file is part of Flutter-Sound.
 *
 * Flutter-Sound is free software: you can redistribute it and/or modify
 * it under the terms of the Mozilla Public License version 2 (MPL-2.0),
 * as published by the Mozilla organization.
 *
 * Flutter-Sound is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * MPL General Public License for more details.
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

// The three interfaces to the platform
// ------------------------------------

/// ------------------------------------------------------------------
/// # The Flutter Sound library
///
/// Flutter Sound is composed with three main modules/classes
/// - [FlutterSoundPlayer]. Everything about the playback functions
/// - [FlutterSoundRecorder]. Everything about the recording functions
/// - [FlutterSoundHelper]. Some utilities to manage audio data.
/// ------------------------------------------------------------------
library;

// The interfaces to the platforms specific implementations
// --------------------------------------------------------
//export 'package:flutter_sound_platform_interface/flutter_sound_platform_interface.dart';

// everything : no documentation
// @nodoc
// library;

export 'package:flutter_sound_platform_interface/flutter_sound_platform_interface.dart';

/// Main
///library tau;
export 'flutter_sound_player.dart';
export 'flutter_sound_recorder.dart';
export 'flutter_sound_helper.dart';
//export 'tau.dart';

import 'dart:typed_data' show Uint8List;
import 'package:logger/logger.dart' show Level, Logger;
import 'flutter_sound_player.dart';
import 'flutter_sound_recorder.dart';

/// For internal code. Do not use.
///
/// The possible states of the players and recorders
/// @nodoc
enum Initialized {
  /// The object has been created but is not initialized
  notInitialized,

  /// The object is initialized and can be fully used
  fullyInitialized,
}

/// The usual file extensions used for each codecs
const List<String> ext = [
  '', // defaultCodec
  '.aac', // aacADTS
  '.opus', // opusOGG
  '.caf', // opusCAF
  '.mp3', // mp3
  '.ogg', // vorbisOGG
  '.pcm', // pcm16
  '.wav', // pcm16WAV
  '.aiff', // pcm16AIFF
  '_pcm.caf', // pcm16CAF
  '.flac', // flac
  '.mp4', // aacMP4
  '.amr', // AMR-NB
  '.amr', // amr-WB
  '.pcm', // pcm8
  '.pcm', // pcmFloat32
  '.pcm', //codec.pcmWebM,
  '.webm', // codec.opusWebM,
  '.webm', // codec.vorbisWebM,
  '.wav', // pcmFloat32WAV
];

/// The valid file extensions for each codecs
const List<List<String>> validExt = [
  [''], // defaultCodec
  ['.aac', '.adt', '.adts'], // aacADTS
  ['.opus', '.ogg'], // opusOGG
  ['.caf'], // opusCAF
  ['.mp3'], // mp3
  ['.ogg'], // vorbisOGG
  ['.pcm', '.aiff'], // pcm16
  ['.wav'], // pcm16WAV
  ['.aiff'], // pcm16AIFF
  ['.caf'], // pcm16CAF
  ['.flac'], // flac
  ['.mp4', '.aac', '.m4a'], // aacMP4
  ['.amr', '.3ga'], // AMR-NB
  ['.amr', '.3ga'], // amr-WB
  ['.pcm', '.aiff'], // pcm8
  ['.pcm', '.aiff'], // pcmFloat32
  ['.pcm', '.webm'], //codec.pcmWebM,
  ['.opus', '.webm'], // codec.opusWebM,
  ['.webm'], // codec.vorbisWebM,
  ['.wav'], // pcmFloat32WAV
];

/// Food is an abstract class which represents objects that can be sent
/// to a player when playing data from astream or received by a recorder
/// when recording to a Dart Stream.
///
@Deprecated('Don\'t use anymore Food, but directely your buffers')
/// @nodoc
abstract class Food {
  /// use internally by Flutter Sound
  Future<void> exec(FlutterSoundPlayer player);

  /// use internally by Flutter Sound
  void dummy(FlutterSoundPlayer player) {} // Just to satisfy `dartanalyzer`
}

/// FoodData are the regular objects received from a recorder when recording to a Dart Stream
/// or sent to a player when playing from a Dart Stream
@Deprecated('Don\'t use anymore Food, but directely your buffers')
/// @nodoc
class FoodData extends Food {
  /// the data to be sent (or received)
  Uint8List? data;

  /// The constructor, specifying the data to be sent or that has been received
  /* ctor */
  FoodData(this.data);

  /// Used internally by Flutter Sound
  @override
  Future<void> exec(FlutterSoundPlayer player) => player.feedFromStream(data!);
}

/// foodEvent is a special kind of food which allows to re-synchronize a stream
/// with a player that play from a Dart Stream
/// @nodoc
@Deprecated('Don\'t use anymore Food, but directely your buffers')
class FoodEvent extends Food {
  /// The callback to fire when this food is synchronized with the player
  Function on;

  /// The constructor, specifying the callback which must be fired when synchronization is done
  /* ctor */
  FoodEvent(this.on);

  /// Used internally by Flutter Sound
  @override
  Future<void> exec(FlutterSoundPlayer player) async => on();
}

/// This is **THE** main Flutter Sound class.
///
/// For future expansion. Do not use.
/// This class is not instanciable. Use the expression [FlutterSound()] when you want to get the Singleton.
///
/// This class is used to access the main functionalities of Flutter Sound. It declares also
/// a default [FlutterSoundPlayer] and a default [FlutterSoundRecorder] that can be used
/// by the App, without having to build such objects themselves.
/// @nodoc
class FlutterSound {
  Logger logger = Logger(level: Level.debug);

  /// The FlutterSound Logger getter
  //Logger get logger => _logger;

  /// The FlutterSound Logger setter
  //set logger(Logger aLogger) {
  //  _logger = aLogger;
  //  // TODO
  // Here we must call flutter_sound_core if necessary
  //}

  // ---------------------------------------------------------------------------------------------------------------------

  /// the static Singleton
  static final FlutterSound _singleton = FlutterSound._internal();

  /// The factory which returns the Singleton
  factory FlutterSound() {
    return _singleton;
  }

  /// Private constructor of the Singleton
  /* ctor */
  FlutterSound._internal();

  /// This instance of [FlutterSoundPlayer] is just to be smart for the App.
  /// The Apps can use this instance without having to create a [FlutterSoundPlayer] themselves.
  FlutterSoundPlayer thePlayer = FlutterSoundPlayer();

  /// TODO
  void internalOpenSessionForRecording() {
    //todo
  }

  // ----------------------------------------------------------------------------------------------------------------------
}
